Classes/Emogrifier.php

```diff
@@ -1289,40 +1292,63 @@ function (array $matches) {
+            if (strstr($trimmedLowercaseSelector, ':not')) {
+                list($trimmedLowercaseSelector, $notSelector) = explode(':not', $trimmedLowercaseSelector);
+            }
 
-            // Advanced selectors are going to require a bit more advanced emogrification.
-            // When we required PHP 5.3, we could do this with closures.
-            $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
-                '/([^\\/]+):nth-child\\(\\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthChild'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
-            $finalXpath = preg_replace_callback(
-                '/([^\\/]+):nth-of-type\\(\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthOfType'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
+            $finalXpath = '//' . $this->translateCssToXpathPass($trimmedLowercaseSelector);
+
+            if ($notSelector) {
+                $notSelector = trim($notSelector, '()');
+                $finalXpath = $finalXpath . '[not(' . $this->translateCssToXpathPass($notSelector, true) . ')]';
+            }
 
             $this->caches[self::CACHE_KEY_SELECTOR][$xPathKey] = $finalXpath;
         }
         return $this->caches[self::CACHE_KEY_SELECTOR][$xPathKey];
     }
 
+    private function translateCssToXpathPass($trimmedLowercaseSelector, $inline = false)
+    {
+        $roughXpath = preg_replace(
+            array_keys($this->xPathRules),
+            $this->xPathRules,
+            $trimmedLowercaseSelector
+        );
+
+        $xPathWithIdAttributeMatchers = preg_replace_callback(
+            self::ID_ATTRIBUTE_MATCHER,
+            [$this, 'matchIdAttributes'],
+            $roughXpath
+        );
+
+        $matchClassAttributesCallback = [$this, 'matchClassAttributes'];
+        if ($inline) {
+            $matchClassAttributesCallback = [$this, 'matchClassAttributesInline'];
+        }
+        $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
+            self::CLASS_ATTRIBUTE_MATCHER,
+            $matchClassAttributesCallback,
+            $xPathWithIdAttributeMatchers
+        );
+
+        // Advanced selectors are going to require a bit more advanced emogrification.
+        // When we required PHP 5.3, we could do this with closures.
+        $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
+            '/([^\\/]+):nth-child\\(\\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
+            [$this, 'translateNthChild'],
+            $xPathWithIdAttributeAndClassMatchers
+        );
+        $finalXpath = preg_replace_callback(
+            '/([^\\/]+):nth-of-type\\(\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
+            [$this, 'translateNthOfType'],
+            $xPathWithIdAttributeAndClassMatchers
+        );
```

Comment:
If I read this correctly, you're also just moving code around. To have things separate and to make this easier to review, could you please split this into two PRs: One refactoring PR that just moves code around without changing the behavior, and one PR after that adds the new behavior, making use of the then-already-added refactorings.

Reply:
Not just moving code around. The same exact code is supposed to be run 2 times.

First for regular xPath query and second time to create :not query. There is no point in writing same code twice. Only thing changed in the code is I added variable $matchClassAttributesCallback that changes if the result should be full xPath selector or only the inline part needed for :not query.

Reply:
Ah, okay. Thanks for the clarification!